Add Makefile functionality to generate a file with all the Chapel environment variables

sungeunchoi opened the pull request:
Generate a header file with #defines representing the Chapel environment variables printed out by the --simple option to printchplenv.  Since the primary use of these is for conditional compilation, they cannot be defined to be strings or non-integer values.  Consequently, the name of the generated #define is the
environment variable name concatenated with '_' and the value in upper case, e.g., CHPL_COMM_SUBSTRATE_UDP, and they have no value.

The header file, named chpl-env-gen.h, is placed in the runtime gen directory.  The path to the header is added to RUNTIME_INCL and runtime code can include the file directly.

There is currently at least one need for such functionality in the marenostrum launch, and it is anticipated that there will be more.

Also add a compile-only C test.

sungeunchoi commented:
@ronawho @mppf 

I'm not in a hurry to get this in, so if it makes sense to delay until next release, that's fine with me also.

ronawho commented on runtime/Makefile:
```diff
@@ -39,6 +42,17 @@ ifneq ($(CHPL_MAKE_LAUNCHER),none)
 
+$(CHPL_ENV_HEADER):
+	echo $(RUNTIME_INCLS)
+	@echo $(CHPL_ENV_HEADER)
```
Not sure if this will fix it, but I think you need to `@mkdir -p $(CHPL_ENV_HEADER_DIR)` before trying to create CHPL_ENV_HEADER

sungeunchoi replied:
Good catch.  Still think we need better error messages from the smoke test.

sungeunchoi commented:
@ronawho 

Finally worked!  Thanks for the tip.

sungeunchoi commented:
@ronawho 

Updated the commit to implemented the changes we discussed as well as a test case.  Also, I added a .cleanfiles even tho they don't work for C tests (yet).

ronawho commented on test/runtime/sungeun/chpl-env-gen.precomp:
```diff
@@ -0,0 +1,23 @@
+#!/usr/bin/env python
```
I think this would be a little cleaner as:

```python
#!/usr/bin/env python

import os
import sys
import subprocess

chpl_def_home = os.path.join(os.path.dirname(__file__), '..', '..', '..')
chpl_home = os.getenv('CHPL_HOME', chpl_def_home)

printchplenv = os.path.join(chpl_home, 'util', 'printchplenv')
printchplenv_cmd = [printchplenv, '--simple']

p = subprocess.Popen(printchplenv_cmd, stdout=subprocess.PIPE).communicate()[0]
chpl_env = dict(map(lambda l: l.split('='), p.splitlines()))

testname = sys.argv[1]
genfile = testname+'.test.gen.c'
with open(genfile, 'w') as f:
    for key,val in chpl_env.items():
        key_val = key+'_'+val.upper()
        f.write('#ifndef %s\n'%(key_val))
        f.write('#error "%s undefined or does not match runtime definition"\n'%(val))
        f.write('#endif\n')

```

sungeunchoi replied:
I like, except `chpl_def_home`, since it makes me feel like I need to maintain that line of code if the directory ever moves.  Is CHPL_HOME ever *not* defined?  If not, I'd just as soon leave off the default value for `getenv`.

ronawho replied:
fair enough. CHPL_HOME should always be defined when running under the testing system 

Maybe just leave it as `chpl_home = os.getenv('CHPL_HOME')`

that way chpl_home will be `None` if $CHPL_HOME isn't set and it'll bark early on rather than trying to run a non-existent printchplenv

ronawho commented:
Other than making the precomp a little more python, looks good!

sungeunchoi commented:
@ronawho Updated the precomp script.

ronawho commented:
:+1:
